### PR TITLE
Ptref: arrivals bug

### PR DESCRIPTION
**Pull request by pbougue**
Arrivals iterated on departures, it is not the case anymore

JIRA: http://jira.canaltp.fr/browse/NAVITIAII-1813

### CODE REVIEW

File: source/routing/next_stop_time.cpp [hunk 1]
```diff
@@ -411,9 +423,14 @@ DateTime get_next_departure(DateTime dt, const type::FrequencyVehicleJourney& fr
     return DateTimeUtils::set(date, DateTimeUtils::hour(lower_bound + x * freq_vj.headway_secs));
 }
 
-DateTime get_previous_arrival(DateTime dt, const type::FrequencyVehicleJourney& freq_vj, const type::StopTime& st, const bool adapted) {
-    const u_int32_t lower_bound = DateTimeUtils::hour(freq_vj.start_time + st.arrival_time);
-    const u_int32_t upper_bound = DateTimeUtils::hour(freq_vj.end_time + st.arrival_time);
+DateTime get_previous_stop_time(const bool is_on_departures,
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: I really think this should be an enum :confounded:

File: source/ptreferential/tests/ptref_test.cpp [hunk 2]
```diff
@@ -261,14 +261,14 @@ BOOST_AUTO_TEST_CASE(physical_modes) {
     b.finish();
 
     auto indexes = make_query(navitia::type::Type_e::Line, "line.uri=A", *(b.data));
-    BOOST_CHECK_EQUAL(indexes.size(), 1);
-    BOOST_CHECK_EQUAL(b.data->pt_data->lines[indexes.front()]->physical_mode_list.size(), 2);
+    BOOST_REQUIRE_EQUAL(indexes.size(), 1);
+    BOOST_REQUIRE_EQUAL(b.data->pt_data->lines[indexes.front()]->physical_mode_list.size(), 2);
```
Comment (TeXitoi) [inline source/ptreferential/tests/ptref_test.cpp]: You have to check the correctness of `indexes.front()` or use `lines.at(indexes.front())`, else a bug will crash or worst, pass.

File: source/ptreferential/tests/ptref_test.cpp [hunk 3]
```diff
@@ -261,14 +261,14 @@ BOOST_AUTO_TEST_CASE(physical_modes) {
...
     BOOST_CHECK_EQUAL(b.data->pt_data->lines[indexes.front()]->physical_mode_list[0]->name, "Car");
     BOOST_CHECK_EQUAL(b.data->pt_data->lines[indexes.front()]->physical_mode_list[1]->name, "Metro");
 
     indexes = make_query(navitia::type::Type_e::Line, "line.uri=C", *(b.data));
-    BOOST_CHECK_EQUAL(indexes.size(), 1);
-    BOOST_CHECK_EQUAL(b.data->pt_data->lines[indexes.front()]->physical_mode_list.size(), 1);
+    BOOST_REQUIRE_EQUAL(indexes.size(), 1);
+    BOOST_REQUIRE_EQUAL(b.data->pt_data->lines[indexes.front()]->physical_mode_list.size(), 1);
```
Comment (TeXitoi) [inline source/ptreferential/tests/ptref_test.cpp]: same here

**Comment by TeXitoi** [comment]
It needs more tests.

File: source/routing/next_stop_time.cpp (hunk 1 above)
Comment (pbougue) [inline source/routing/next_stop_time.cpp]: I really think so too (maybe more a struct for both prev/next and arrival/departure).
And it's an open-source project, as you taught me.

File: source/time_tables/previous_passages.cpp [hunk 4]
```diff
@@ -59,25 +59,28 @@ previous_passages(const std::string &request,
     std::remove_if(handler.journey_pattern_points.begin(),
                    handler.journey_pattern_points.end(), vis.predicate);
 
-    auto passages_dt_st = get_stop_times(handler.journey_pattern_points,
-                            handler.date_time, handler.max_datetime,
-                            nb_stoptimes, data, disruption_active, accessibilite_params,
-                            false);
+    const bool is_on_departures(vis.api_pb == pbnavitia::PREVIOUS_DEPARTURES);
```
Comment (xlqian) [inline source/time_tables/previous_passages.cpp]: why not `const bool is_on_departures = vis.api_pb == pbnavitia::PREVIOUS_DEPARTURES;`

File: source/routing/next_stop_time.cpp (hunk 1 above)
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: too bad you still are in the training period :wave:

File: source/time_tables/previous_passages.cpp (hunk 4 above)
Comment (pbougue) [inline source/time_tables/previous_passages.cpp]: looks clearer to me that way :-/

File: source/routing/next_stop_time.cpp (hunk 1 above)
Comment (pbougue) [inline source/routing/next_stop_time.cpp]: Yep, cause I still don't know how to make this `fu` symbol ;)

File: source/time_tables/previous_passages.cpp (hunk 4 above)
Comment (antoine-de) [inline source/time_tables/previous_passages.cpp]: I do aggry with @pbougue, I find the `()` a bit clearer there, else there is too much `=` :wink:

File: source/routing/next_stop_time.cpp (hunk 1 above)
Comment (xlqian) [inline source/routing/next_stop_time.cpp]: Ok for me :)

File: source/time_tables/previous_passages.cpp (hunk 4 above)
Comment (TeXitoi) [inline source/time_tables/previous_passages.cpp]: I prefer with `=`. I vote `=` > `{}` > `()`

File: source/routing/next_stop_time.cpp [hunk 5]
```diff
@@ -226,7 +229,8 @@ previous_valid_frequency_drop_off(const type::JourneyPatternPoint* jpp,
                 continue;
             }
 
-            const auto previous_dt = get_previous_arrival(previous_date, *freq_vj, st, adapted);
+            const auto previous_dt = get_previous_stop_time(is_on_departures, previous_date, *freq_vj,
+                                                          st, adapted);
```
Comment (TeXitoi) [inline source/routing/next_stop_time.cpp]: strange indentation

File: source/routing/next_stop_time.h [hunk 6]
```diff
@@ -49,40 +50,57 @@ namespace routing {
...
         template<typename T> inline bool
         operator()(const T& lhs, const T& rhs) const { return lhs < rhs; }
         DateTime get_time(const type::StopTime& st) const;
         bool is_valid(const type::StopTime& st) const;
     };
-    struct Backward {
+    struct Arrival {
         template<typename T> inline bool
-        operator()(const T& lhs, const T& rhs) const { return lhs > rhs; }
+        operator()(const T& lhs, const T& rhs) const { return lhs < rhs; }
```
Comment (TeXitoi) [inline source/routing/next_stop_time.h]: Is this method needed now?

File: source/routing/raptor.cpp [hunk 7]
```diff
@@ -593,8 +593,8 @@ void RAPTOR::raptor_loop(Visitor visitor,
                     if (prec_labels.transfer_is_initialized(jpp.sp_idx) &&
                         (!is_onboard || visitor.better_or_equal(previous_dt, workingDt, *it_st))) {
                         const auto tmp_st_dt = next_st.next_stop_time(
-                            jpp.idx, previous_dt, visitor.clockwise(), disruption_active,
-                            accessibilite_params.vehicle_properties, jpp.has_freq);
+                            visitor.clockwise(), jpp.idx, previous_dt, visitor.clockwise(),
```
Comment (TeXitoi) [inline source/routing/raptor.cpp]: These 2 `visitor.clockwise()` are really unreadable :(

File: source/time_tables/previous_passages.cpp [hunk 8]
```diff
@@ -59,25 +59,28 @@ previous_passages(const std::string &request,
...
+        } else {
+            passage = handler.pb_response.add_next_arrivals();
         }
         pt::time_period action_period(navitia::to_posix_time(dt_stop_time.first, data), pt::seconds(1));
         auto departure_date = navitia::to_posix_timestamp(dt_stop_time.first, data);
         auto arrival_date = navitia::to_posix_timestamp(dt_stop_time.first, data);
         passage->mutable_stop_date_time()->set_departure_date_time(departure_date);
         passage->mutable_stop_date_time()->set_arrival_date_time(arrival_date);
+
+        fill_pb_object(dt_stop_time.second, data, passage->mutable_stop_date_time()->mutable_properties(), 0, current_datetime, action_period);
```
Comment (TeXitoi) [inline source/time_tables/previous_passages.cpp]: long line

File: source/time_tables/tests/get_stop_times.cpp [hunk 9]
```diff
@@ -345,3 +348,65 @@ BOOST_AUTO_TEST_CASE(test_frequency_over_midnight_for_calendar) {
...
+    b.vj("A")("stop1", "8:00"_t, "8:01"_t)("stop2", "9:00"_t, "9:01"_t)("stop3", "10:00"_t, "10:01"_t);
+    b.vj("B")("stop2", "8:30"_t, "8:31"_t)("stop3", "9:30"_t, "9:31"_t)("stop4", "10:30"_t, "10:31"_t);
+    b.finish();
+    b.data->pt_data->index();
+    b.data->build_uri();
+    b.data->build_raptor();
+
+    auto vec_jpp_stop1 = b.data->pt_data->stop_points_map["stop1"]->journey_pattern_point_list;
+    std::vector<navitia::type::idx_t> jpps_stop1;
+    for(auto jpp : vec_jpp_stop1)
```
Comment (antoine-de) [inline source/time_tables/tests/get_stop_times.cpp]: {}

File: source/time_tables/tests/get_stop_times.cpp [hunk 10]
```diff
@@ -345,3 +348,65 @@ BOOST_AUTO_TEST_CASE(test_frequency_over_midnight_for_calendar) {
...
+BOOST_AUTO_TEST_CASE(test_discrete_next_arrivals_prev_departures){
+    ed::builder b("20150907");
+    b.vj("A")("stop1", "8:00"_t, "8:01"_t)("stop2", "9:00"_t, "9:01"_t)("stop3", "10:00"_t, "10:01"_t);
+    b.vj("B")("stop2", "8:30"_t, "8:31"_t)("stop3", "9:30"_t, "9:31"_t)("stop4", "10:30"_t, "10:31"_t);
+    b.finish();
+    b.data->pt_data->index();
+    b.data->build_uri();
+    b.data->build_raptor();
+
+    auto vec_jpp_stop1 = b.data->pt_data->stop_points_map["stop1"]->journey_pattern_point_list;
```
Comment (antoine-de) [inline source/time_tables/tests/get_stop_times.cpp]: I think you need a function of a lambda for that

File: source/time_tables/tests/get_stop_times.cpp [hunk 11]
```diff
@@ -345,3 +348,65 @@ BOOST_AUTO_TEST_CASE(test_frequency_over_midnight_for_calendar) {
...
+    std::vector<navitia::type::idx_t> jpps_stop1;
+    for(auto jpp : vec_jpp_stop1)
+        jpps_stop1.push_back(jpp->idx);
+    auto result_next_arrivals = get_stop_times(false, jpps_stop1, navitia::DateTimeUtils::min,
```
Comment (antoine-de) [inline source/time_tables/tests/get_stop_times.cpp]: you need to add comments on this test to ease the reading

File: source/time_tables/tests/get_stop_times.cpp [hunk 12]
```diff
@@ -345,3 +348,65 @@ BOOST_AUTO_TEST_CASE(test_frequency_over_midnight_for_calendar) {
...
+                                               navitia::DateTimeUtils::set(1, 0), 100, *b.data, false);
+    BOOST_CHECK_EQUAL(result_next_arrivals.size(), 0);
+    auto result_prev_departures = get_stop_times(true, jpps_stop1, navitia::DateTimeUtils::set(1, 0),
+                                                 navitia::DateTimeUtils::min, 100, *b.data, false);
+    BOOST_REQUIRE_EQUAL(result_prev_departures.size(), 1);
+    BOOST_CHECK_EQUAL(result_prev_departures.at(0).first, "8:01"_t);
```
Comment (antoine-de) [inline source/time_tables/tests/get_stop_times.cpp:390]: wouldn't it be clearer with a BOOST_CHECK_EQUAL_COLLECTION ?

Comment (pbougue) [inline source/time_tables/tests/get_stop_times.cpp:390]: As I only check first, I'll let it that way

**Comment by pbougue** [comment]
Tests added in get_stop_times_test.cpp

**Pull request merged**